platform-sdk/swirlds-common/src/main/java/com/swirlds/common/threading/CountUpLatch.java

```diff
@@ -0,0 +1,187 @@
+     *
+     * @param count the new count, must be greater than the current value
+     * @throws IllegalArgumentException if the new count is less than the current count (the state of this object is
+     *                                  messed up if this happens -- this is an unrecoverable error)
+     */
+    public void set(final long count) {
+        final long result = currentCount.accumulateAndGet(count, SET_COUNT);
+        if (result != count) {
+            throw new IllegalArgumentException(
+                    "Can't set the count to a lower value. Previous = " + result + ", provided = " + count);
+        }
+        lock.lock();
+        try {
+            condition.signalAll();
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    /**
+     * Increment the count by the given delta.
+     * <p>
+     * Methods that update the count are not mutually thread safe. Concurrent threads should never attempt to update
+     * the count. It is, however, safe for many threads to concurrently call methods that query the count.
+     *
+     * @param delta the amount to increment the count by
+     * @throws IllegalArgumentException if the delta is negative
+     */
+    public void add(final long delta) {
+        if (delta < 0) {
+            throw new IllegalArgumentException("Cannot add a negative delta (" + delta + ")");
+        }
+        currentCount.addAndGet(delta);
+
+        lock.lock();
+        try {
+            condition.signalAll();
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    /**
+     * Wait until the count reaches the given value.
+     *
+     * @param count the count to wait for
+     * @throws InterruptedException if interrupted while waiting
+     */
+    public void await(final long count) throws InterruptedException {
+        if (currentCount.get() >= count) {
+            return;
+        }
+
+        while (true) {
+            lock.lock();
+            try {
+                if (currentCount.get() >= count) {
+                    return;
+                }
+                condition.await();
+            } finally {
+                lock.unlock();
+            }
+        }
+    }
+
+    /**
+     * Wait until the count reaches the given value.
+     *
+     * @param count      the count to wait for
+     * @param timeToWait the maximum time to wait
+     * @return true if the count reached the given value, false if the time to wait expired
+     * @throws InterruptedException if interrupted while waiting
+     */
+    @SuppressWarnings("ResultOfMethodCallIgnored")
+    public boolean await(final long count, final Duration timeToWait) throws InterruptedException {
+        if (currentCount.get() >= count) {
+            return true;
+        }
+
+        final Instant start = Instant.now();
+
+        while (true) {
+            final Instant now = Instant.now();
+            final Duration elapsed = Duration.between(start, now);
```

Comment:
I think I get what you're doing, you want a maximum total duration, while still handling potentially many wakeups between start and end.
Packing this in a method `remainingTime = checkMaximumTimeout(start, timeToWait)` might make it clearer. an `if(remainingTime > 0)...else return (value)` block after that could handle the wait-or-return part.

Reply:
I've pulled the computation for remaining time into a helper method.